engine/scheduler.cpp

```diff
@@ -865,6 +971,37 @@ struct engine::scheduler::scheduler_handle::impl : utils::noncopyable {
         return tests_data;
     }
 
+    /// Finds any pending exec_datas that correspond to tests needing execenv
+    /// cleanup.
+    ///
+    /// \return The collection of test_exec_data objects that have their
+    /// specific execenv property set.
+    test_exec_data_vector
+    tests_needing_execenv_cleanup(void)
+    {
+        test_exec_data_vector tests_data;
+
+        for (exec_data_map::const_iterator iter = all_exec_data.begin();
+             iter != all_exec_data.end(); ++iter) {
+            const exec_data_ptr data = (*iter).second;
+
+            try {
+                test_exec_data* test_data = &dynamic_cast< test_exec_data& >(
+                    *data.get());
```

Comment:
```suggestion
                test_exec_data* test_data = &dynamic_cast< test_exec_data& >(
                    *data.get());
```
`dynamic_cast` can fail here. The code needs to defensively check for a `nullptr` (even if it's using `INV`).

Reply:
IIRC, `dynamic_cast` returns `nullptr` in a case when target type is a pointer type, in our case with a reference type as a target it throws an exception instead. And there is a `catch` for that. Actually, this part of the code is similar to the existing logic, for instance: https://github.com/freebsd/kyua/blob/2346346090062e2e52d5356c7a546fbc89829118/engine/scheduler.cpp#L853